airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/source/relationaldb/state/CursorStateMessageProducer.java

```diff
@@ -0,0 +1,144 @@
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CursorStateMessageProducer.class);
+
+  private final StateManager stateManager;
+  private final String initialCursor;
+  private String currentMaxCursor;
+  private long currentMaxCursorRecordCount = 0L;
+
+  // We keep this field just to control logging frequency.
+  private int totalRecordCount = 0;
+  private AirbyteStateMessage intermediateStateMessage = null;
+
+  private boolean cursorOutOrOrderDetected = false;
+
+  public CursorStateMessageProducer(final StateManager stateManager,
+                                    final String initialCursor) {
+    this.stateManager = stateManager;
+    this.initialCursor = initialCursor;
+    this.currentMaxCursor = initialCursor;
+  }
+
+  /**
+   * @return
+   */
+  @Override
+  public AirbyteStateMessage generateStateMessageAtCheckpoint(final ConfiguredAirbyteStream stream) {
+    // At this stage intermediate state message should never be null; otherwise it would have been
+    // blocked by shouldEmitStateMessage check.
+    final AirbyteStateMessage message = intermediateStateMessage;
+    intermediateStateMessage = null;
+    if (cursorOutOrOrderDetected) {
+      LOGGER.warn("Intermediate state emission feature requires records to be processed in order according to the cursor value. Otherwise, "
+          + "data loss can occur.");
+    }
+    return message;
+  }
+
+  /**
+   * Note: We do not try to catch exception here. If error/exception happens, we should fail the sync,
+   * and since we have saved state message before, we should be able to resume it in next sync if we
+   * have fixed the underlying issue, of if the issue is transient.
+   *
+   * @param message
+   * @return
+   */
+  @Override
+  public AirbyteMessage processRecordMessage(final ConfiguredAirbyteStream stream, AirbyteMessage message) {
+    totalRecordCount++;
+    final String cursorField = IncrementalUtils.getCursorField(stream);
+    if (message.getRecord().getData().hasNonNull(cursorField)) {
+      final String cursorCandidate = getCursorCandidate(cursorField, message);
+      final JsonSchemaPrimitive cursorType = IncrementalUtils.getCursorType(stream,
+          cursorField);
+      final int cursorComparison = IncrementalUtils.compareCursors(currentMaxCursor, cursorCandidate, cursorType);
+      if (cursorComparison < 0) {
+        // Update the current max cursor only when current max cursor < cursor candidate from the message
+        if (!Objects.equals(currentMaxCursor, initialCursor)) {
+          // Only create an intermediate state when it is not the first record.
+          intermediateStateMessage = createStateMessage(stream, totalRecordCount);
+        }
+        currentMaxCursor = cursorCandidate;
+        currentMaxCursorRecordCount = 1L;
+      } else if (cursorComparison == 0) {
+        currentMaxCursorRecordCount++;
+      } else if (cursorComparison > 0) {
+        cursorOutOrOrderDetected = true;
```

Comment:
Should we do something to `currentMaxCursorRecordCount` here too?

Reply:
I realized we do not need this variable, as the count is handled by the iterator itself. Removing this for all. I need more tests on this :) adopting one connector seems to be a great idea to do so.